File: 687.Longest-Univalue-Path.go
// https://leetcode.com/problems/longest-univalue-path/
// algorithms
// Easy (33.36%)
// Total Accepted:    51,984
// Total Submissions: 155,850
// beats 100.0% of golang submissions

package leetcode

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

var res int

func longestUnivaluePath(root *TreeNode) int {
	if root == nil {
		return 0
	}

	res = 0

	recursive(root)

	return res
}

func recursive(node *TreeNode) int {
	if node == nil {
		return 0
	}

	l, r := recursive(node.Left), recursive(node.Right)

	if node.Left != nil && node.Val == node.Left.Val {
		l++
	} else {
		l = 0
	}
	if node.Right != nil && node.Val == node.Right.Val {
		r++
	} else {
		r = 0
	}

	res = max(res, l+r)

	return max(l, r)
}

func max(a, b int) int {
	if a > b {
		return a
	}

	return b
}

File: 687.Longest-Univalue-Path.java
// https://leetcode.com/problems/max-area-of-island/
//
// algorithms
// Easy (34.58%)
// Total Accepted: 68,755
// Total Submissions: 198,827


/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode(int x) { val = x; } }
 */
class Solution {
    static int res;

    public int longestUnivaluePath(TreeNode root) {
        res = 1;
        recursive(root);

        return res - 1;
    }

    public int recursive(TreeNode node) {
        if (node == null) {
            return 0;
        }

        int left = recursive(node.left);
        int right = recursive(node.right);
        int resTmp = 1;
        if (node.left != null && node.left.val == node.val) {
            resTmp += left;
            left++;
        } else {
            left = 0;
        }
        if (node.right != null && node.right.val == node.val) {
            resTmp += right;
            right++;
        } else {
            right = 0;
        }
        res = Math.max(resTmp, res);

        return Math.max(1, Math.max(left, right));
    }
}